lib/TPP/Transforms/Utils/VNNIUtils.cpp

```diff
@@ -20,22 +21,54 @@ namespace mlir {
 namespace vnni {
 namespace utils {
 
-std::optional<int64_t> getVnniBlockingFactor(Type type) {
+std::optional<int64_t> getVnniBlockingFactor(Type type, Operation *op) {
   auto elementType = getElementTypeOrSelf(type);
-  if (elementType.isBF16())
+  if (elementType.isBF16()) {
+    // Check if a VNNI factor hint is associated to the IR via DLTI.
+    auto deriveVnniFromDLTI = [&]() -> std::optional<int64_t> {
+      if (!op)
+        return std::nullopt;
+      ModuleOp moduleOp = op->getParentOfType<mlir::ModuleOp>();
+      if (!moduleOp)
+        return std::nullopt;
+      TargetSystemSpecInterface sysSpec = moduleOp.getTargetSystemSpec();
+      if (!sysSpec)
+        return std::nullopt;
+      auto deviceId = StringAttr::get(moduleOp->getContext(), "CPU");
+      auto deviceSpec = sysSpec.getDeviceSpecForDeviceID(deviceId);
+      if (!deviceSpec)
+        return std::nullopt;
+      auto vnniId = StringAttr::get(moduleOp->getContext(), "vnni");
+      DataLayoutEntryInterface entry =
+          (*deviceSpec).getSpecForIdentifier(vnniId);
+      if (!entry)
+        return std::nullopt;
+      Attribute value = entry.getValue();
+      if (auto intAttr = llvm::dyn_cast<IntegerAttr>(value))
+        return intAttr.getInt();
+      return std::nullopt;
+    };
+    if (auto vnniFactor = deriveVnniFromDLTI())
+      return *vnniFactor;
+
     return libxsmm_cpuid_dot_pack_factor(LIBXSMM_DATATYPE_BF16);
+  }
   return std::nullopt;
```

Comment:
Shouldn't we return `1`  when there is no blocking factor? This would fix the issue below when we treat "no blocking" and "don't know" as the same answer.

Reply:
Factor of `1` would imply you still want 3D shape like `<8x16x1>`.
We could of course choose to do nothing if it's `1` but I see no reason to refactor everything right now.

Reply:
Could also be `0`. What I mean is that we're using `nullopt` to mean both states and that breaks assumptions on the user side.

Reply:
Actually I could add some verification on the internal queries to make sure the value is reasonable.
But the way I see it, I ask for a blocking factor and if I get it, it means I'm good to continue.
If not, then it's up to me as a user to decide what I expect to do.

Reply:
Than don't use `std::optional`. Optional created a ternary state: { don't have it, have null, have value }. Here the usage conflates the first two, but the code that uses is is still required to make the difference.

So, either use `0` as the null set, or actually make use of `optional` and separate the three states.

Reply:
After more offline discussions, I agree.
Refactored API to return `unsigned` and added more checks through the codebase.